### DISCUSSION TITLE
输入双拼怎么同时显示全拼和原拼？

### DISCUSSION THREAD

**Discussion by hegotit** [rime/home#1466] [Q&A]
比如输入键位"code"，在输入码区显示"cuode, code"，方便双拼和英文混打

**Comment by rightester** [comment]
我预设你码表里的项是全拼形式`cuode 错的`而不是双拼形式`code 错的` （这很重要，影响如何实现）
这样的话你可能需要一个lua_processor，顺序在speller前，功能上添加一个prompt给最后一个segment，prompt的内容就记录之前每次击键的字母，拼接起来。
这样子实现的效果看起来比较像是这样的（prompt整个区域的字符串都是可以自定义的）
[![ prompt 的示意图 ](https://user-images.githubusercontent.com/18041500/190980054-7e944f5f-a381-4c73-ad6a-254a00c09e44.png)](https://github.com/hchunhui/librime-lua/wiki/Scripting#Segment)

（如果你需要，我可以之后贴出一段代码和使用方式供参考）

**Reply by hegotit** [reply]
码表是全拼形式，双拼依赖于全拼码表。
方便的话可以给个demo参考下吗？

**Comment by gaboolic** [comment]
也可以在dict里面加上code code

**Comment by rightester** [comment]
今天我再次翻看提问内容时我发现了我的一些疏漏；如果你想要实现的是双拼英文混打功能，那我前述提及方式并不能实现这种效果，因为它只是一个纯显示用途的提示，我错误理解成了你想要一种全拼转双拼的某种提示……
如果你的中英混输词是已知且静态的，你可以直接将它们加入自定义短语的词典；
不过我猜你想实现的是输入时动态产生双拼中文和英文混输词？如果是这样，我提出一种办法，仅供参考：你需要的可能是lua_translator和一个英文词典，能将特定的序列转换成英文并产生候选……
例如这样：输入nqb，正常只会出现牛逼；然后左右键移动光标，translator通过segment判断当前光标是否移动了（是否seg不完整），移动了则将当前输入码作为英文直接产生一个候选项；当光标移动至输入码只有1码时，分别产生大小写两种候选；任意时候发现输入码匹配英文词典里的一个英文单词，则同时产生这个单词的候选项。
这样应该能实现一部分你需求的功能大概，欢迎讨论 :)
（不直接在任意时刻都产生混英候选是因为排列组合可能性太多了，没有智能算法组合和过滤效果实在是不堪设想）

**Reply by rightester** [reply]
在群里发起话题小小讨论了一下，似乎没有很好的解决办法，最好和最常用的办法就是加进词库
用群友的话说就是“反正这话题车了不止5次了”
要我说的话，这算是Rime架构设计上的潜在缺陷导致的问题，且由于需求较低，没有直接受益开发者用户，所以也没人开发。
就lua提供的接口来说，理论上有一定可行度。

**Reply by hegotit** [reply]
之前就是想要一个纯显示用途的提示就够了，这样打英文的时候能知道自己打的是什么，觉得OK就可以直接Enter，不过目前用了雾凇拼音，可以实现中英混打了

**Comment by lotem** [comment]
不需要大費周章。
拼寫運算支持改編顯示的輸入碼。
如果不配置任何轉寫規則，那麼輸入碼也就是雙拼碼會原樣顯示。
rime/rime-double-pinyin 輸入方案配置爲顯示全拼。
當然也可以顯示兩者，不過轉寫規則要寫得複雜一些。
大致思路是先把雙拼輸入碼[複寫](https://github.com/rime/rime-cangjie/blob/75b10325bf4c590a3ffef7039a6f052a729edc55/cangjie5.schema.yaml#L68)一遍，再將其中之一轉寫爲全拼。

**Comment by byzod** [comment]
> 大致思路是先把雙拼輸入碼[複寫](https://github.com/rime/rime-cangjie/blob/75b10325bf4c590a3ffef7039a6f052a729edc55/cangjie5.schema.yaml#L68)一遍，再將其中之一轉寫爲全拼。

感谢指点，但是遇到了两个问题

1. 辅码颜色不能单独定义，但实际上只是起到参考作用，颜色太明显会分散注意力。不知道能否做到像qq输入法这样，辅码单独改为浅色，仅供参考
2. 其次，由于辅码显示是简单的regex替换，所以对于简拼码，是没有办法获取结果实际的完整读音的 （qq是从候选中获取读音`qing`，而rime简单的替换无法获取该信息）
![2](https://github.com/rime/home/assets/848607/e7649ad9-6e68-451c-b93d-759bb810bc39)


4. [根据文档说明](https://github.com/rime/home/wiki/SpellingAlgebra#%E6%8B%BC%E5%AF%AB%E9%81%8B%E7%AE%97%E7%9A%84%E9%81%8B%E7%AE%97%E5%AD%90)，xform算子是不会用utf32来运算的，所以当替换时的字符包含unicode符号时就会乱码
![3](https://github.com/rime/home/assets/848607/4fef80ef-12a3-4199-b3be-a7e45480a741)
![4](https://github.com/rime/home/assets/848607/7221dcf9-6bf2-42b7-bfd6-d8d1de089e39)



而之所以要用这个，是因为双拼码和全拼码需要一个大写一个小写来分开处理，而为了保留小写码中本来就是大写的码，只能借助中间字符，而中间字符不能用unicode没办法就只能从ascii里找来一堆控制码，甚至不够用还要数字来凑😅

如果有办法让regex用除了`\U`和`\L`以外的映射，或者有办法让`xform`识别ascii以外的字符，就能解决这个问题勒，不知道大家有什么好方法没有？

附：用在`double_pinyin_flypy.custom.yaml`里的配置，其他双拼理论上把中间那段从相应yaml里的`translator/preedit_format`抄过来就行

```yaml

  # 输入时显示双拼编码+全拼码
  translator/preedit_format:
    - xform/(.+)/$1 \($1\)/ # 复制一份副本
    # - xlit/ABCDEFGHIJKLMNOPQRSTUVWXYZ/𝐀𝐁𝐂𝐃𝐄𝐅𝐆𝐇𝐈𝐉𝐊𝐋𝐌𝐍𝐎𝐏𝐐𝐑𝐒𝐓𝐔𝐕𝐖𝐗𝐘𝐙/ # 保留原大写
    - xlit/ABCDEFGHIJKLMNOPQRSTUVWXYZ/1234/ # 保留原大写
    - xform/(.+)(?= \()/\U$1\E/ # 保留原码的转为大写
    
    - xform/([bpmfdtnljqx])n/$1ijao/ # 从原double_pinyin_flypy.schema.yaml复制的内容 - start
    - xform/(\w)g/$1eng/
    - xform/(\w)q/$1iu/
    - xform/(\w)w/$1ei/
    - xform/([dtnlgkhjqxyvuirzcs])r/$1uan/
    - xform/(\w)t/$1ve/
    - xform/(\w)y/$1un/
    - xform/([dtnlgkhvuirzcs])o/$1uo/
    - xform/(\w)p/$1ie/
    - xform/([jqx])s/$1iong/
    - xform/(\w)s/$1ong/
    - xform/(\w)d/$1ai/
    - xform/(\w)f/$1en/
    - xform/(\w)h/$1ang/
    - xform/(\w)j/$1an/
    - xform/([gkhvuirzcs])k/$1uai/
    - xform/(\w)k/$1ing/
    - xform/([jqxnl])l/$1iang/
    - xform/(\w)l/$1uang/
    - xform/(\w)z/$1ou/
    - xform/([gkhvuirzcs])x/$1ua/
    - xform/(\w)x/$1ia/
    - xform/(\w)c/$1ao/
    - xform/([dtgkhvuirzcs])v/$1ui/
    - xform/(\w)b/$1in/
    - xform/(\w)m/$1ian/
    - xform/([aoe])\1(\w)/$1$2/
    - "xform/(^|[ '])v/$1zh/"
    - "xform/(^|[ '])i/$1ch/"
    - "xform/(^|[ '])u/$1sh/"
    - xform/([jqxy])v/$1u/
    - xform/([nl])v/$1ü/
    - xform/ü/v/  # ü 显示为 v  # 从原double_pinyin_flypy.schema.yaml复制的内容 -  end
    
    - xform/(.*)/\L$1/ # 大写副本改回小写
    # - xlit/𝐀𝐁𝐂𝐃𝐄𝐅𝐆𝐇𝐈𝐉𝐊𝐋𝐌𝐍𝐎𝐏𝐐𝐑𝐒𝐓𝐔𝐕𝐖𝐗𝐘𝐙/ABCDEFGHIJKLMNOPQRSTUVWXYZ/ # 恢复原大写
    - xlit/1234/ABCDEFGHIJKLMNOPQRSTUVWXYZ/ # 恢复原大写
    - xform/ (?!\()/'/ # 空格改 '
```

**Reply by hegotit** [reply]
```yaml
    - xform/(.+)/$1 \($1\)/  # 复制一份副本
    - xlit/QWERTYUIOPASDFGHJKLZXCVBNM/ⓆⓌⒺⓇⓉⓎⓊⒾⓄⓅⒶⓈⒹⒻⒼⒽⒿⓀⓁⓏⓍⒸⓋⒷⓃⓂ  # 保留原大写
    - xform/(.+)(?<= \()(.+)/$1\U$2\E/ # 保留原码的转为大写
    # ...
    - xform/(.*)/\L$1/ # 大写副本改回小写
    - xlit/ⓆⓌⒺⓇⓉⓎⓊⒾⓄⓅⒶⓈⒹⒻⒼⒽⒿⓀⓁⓏⓍⒸⓋⒷⓃⓂ/QWERTYUIOPASDFGHJKLZXCVBNM/ # 恢复原大写
```
![image](https://github.com/rime/home/assets/153287866/763153b2-7326-47bd-a303-98ea397abaa2)

多谢参考，我在其基础上改了一版，前面是全拼，括号中是原拼，不过有没有办法用正则去掉括号内的所有空格？

**Reply by byzod** [reply]
只改括号后的有点难吧，前面的可以用`\(`定位，后面的也行，中间的没办法吧

只改括号外的嗦不定有点可能，不知道这套regex系统支不支持那个遇到某个匹配后重置的语法，遇到括号就停止

话说这个带圈字母不也不是ascii吗，为社么xform又能用了啊🤔

**Reply by hegotit** [reply]
如果括号内空格无法清除，那这个修改就失去意义了，本来就是为了直出字符的情况，残留空格会影响辨认。

带圈字母是参考rime-ice写的，实测正常。另外，C++用的UTF-8和其它语言不一致吗？如果一致的话，那[这个文档](https://github.com/rime/home/wiki/SpellingAlgebra#%E6%8B%BC%E5%AF%AB%E9%81%8B%E7%AE%97%E7%9A%84%E9%81%8B%E7%AE%97%E5%AD%90)的描述可能有点误导性：

`「轉寫」是拼寫運算中目前唯一一則將運算元和參數作UTF-32編碼、而非UTF-8編碼處理的運算。 意味着，字母表可以採用ASCII範圍以外的字符、字母表的長度按照Unicode字符數來計算。`

**Reply by byzod** [reply]
你可以考虑前面引入第二层转码，比方说括号里的用`αβγ...χψω`替换，这样就能用`xform/([α-ω]+)\s/$1/`来定位替换了 （只是举个例子，希腊字母没有26个🤭

**Reply by byzod** [reply]
另外如果只是要原码显示用于打英文的话，可以看看[rime-ice](https://github.com/iDvel/rime-ice)的解决方案。[中英混输](https://github.com/iDvel/rime-ice/blob/fc2a5eb7aafa73148b7bc4508e9995b492cecd5a/double_pinyin_flypy.schema.yaml#L27)+歧义项[单独处理](https://github.com/iDvel/rime-ice/blob/fc2a5eb7aafa73148b7bc4508e9995b492cecd5a/double_pinyin_flypy.schema.yaml#L114)，实际用下来感觉不错

![图片](https://github.com/rime/home/assets/848607/62944d83-2fd2-4514-9b84-20da8b824276)

~刚发现你用的就是rimeice啊，当我没说~

**Reply by hegotit** [reply]
希腊字母也不属于ASCII范畴吧！所以还是那个问题，xform到底能不能用ascii范围以外的字符？如果能用的话，那转写字符可选范围就大了，日文假名、韩语谚文甚至注音符号都可以用

**Reply by byzod** [reply]
看了下𝐀是f0 9d 90 80  ，Ⓐ是e2 92 b6  ，大概是只能用第0平面的吧

**Reply by hegotit** [reply]
我发起这个讨论的时候还没有用rime-ice，现在这个特性属于是鸡肋了，不过如果能实现当然更好

**Reply by hegotit** [reply]
> 你可以考虑前面引入第二层转码，比方说括号里的用`αβγ...χψω`替换，这样就能用`xform/([α-ω]+)\s/$1/`来定位替换了 （只是举个例子，希腊字母没有26个🤭

引入了两层转码，但是输入第50个字符的时候，就开始异常了:

![image](https://github.com/rime/home/assets/153287866/36bb6dad-8ea6-415c-996b-224bcf115bd3)

少于50个的时候还是正常的：

![image](https://github.com/rime/home/assets/153287866/5147ea97-82ad-4972-8595-336c4390fa11)

```
  preedit_format:
    - xform/(.+)/$1 \($1\)/  # 复制一份副本
    - xlit/QWERTYUIOPASDFGHJKLZXCVBNM/ⓆⓌⒺⓇⓉⓎⓊⒾⓄⓅⒶⓈⒹⒻⒼⒽⒿⓀⓁⓏⓍⒸⓋⒷⓃⓂ  # 保留原大写
    - xform/(.+)(?<= \()(.+)/$1\U$2\E/ # 保留原码的转为大写
    - xlit/ABCDEFGHIJKLMNOPQRSTUVWXYZ/ⓐⓑⓒⓓⓔⓕⓖⓗⓘⓙⓚⓛⓜⓝⓞⓟⓠⓡⓢⓣⓤⓥⓦⓧⓨⓩ # 保留原码的部分，再从大写转写
    # 改造前start
    - xform/([bpmfdtnljqx])n/$1iao/
    - xform/(\w)g/$1eng/
    - xform/(\w)q/$1iu/
    - xform/(\w)w/$1ei/
    - xform/([dtnlgkhjqxyvuirzcs])r/$1uan/
    - xform/(\w)t/$1ve/
    - xform/(\w)y/$1un/
    - xform/([dtnlgkhvuirzcs])o/$1uo/
    - xform/(\w)p/$1ie/
    - xform/([jqx])s/$1iong/
    - xform/(\w)s/$1ong/
    - xform/(\w)d/$1ai/
    - xform/(\w)f/$1en/
    - xform/(\w)h/$1ang/
    - xform/(\w)j/$1an/
    - xform/([gkhvuirzcs])k/$1uai/
    - xform/(\w)k/$1ing/
    - xform/([jqxnl])l/$1iang/
    - xform/(\w)l/$1uang/
    - xform/(\w)z/$1ou/
    - xform/([gkhvuirzcs])x/$1ua/
    - xform/(\w)x/$1ia/
    - xform/(\w)c/$1ao/
    - xform/([dtgkhvuirzcs])v/$1ui/
    - xform/(\w)b/$1in/
    - xform/(\w)m/$1ian/
    - xform/([aoe])\1(\w)/$1$2/
    - "xform/(^|[ '])v/$1zh/"
    - "xform/(^|[ '])i/$1ch/"
    - "xform/(^|[ '])u/$1sh/"
    - xform/([jqxy])v/$1u/
    - xform/([nl])v/$1ü/
    - xform/ü/v/  # ü 显示为 v
    # 改造前end
    - xform/([ⓐ-ⓩ]+)\s/$1/ #去除紧随其后的制表符
    - xlit/ⓐⓑⓒⓓⓔⓕⓖⓗⓘⓙⓚⓛⓜⓝⓞⓟⓠⓡⓢⓣⓤⓥⓦⓧⓨⓩ/abcdefghijklmnopqrstuvwxyz # 转回小写
    - xlit/ⓆⓌⒺⓇⓉⓎⓊⒾⓄⓅⒶⓈⒹⒻⒼⒽⒿⓀⓁⓏⓍⒸⓋⒷⓃⓂ/QWERTYUIOPASDFGHJKLZXCVBNM/ # 恢复原大写
```

**Comment by lotem** [comment]
鼠鬚管實現了把一部分提示放在光標之後的機制。光標前後的顏色好像是不一樣的，不過默認的 inline preedit 樣式，都顯示在文本框裏，也就無所謂顏色了。

恕我直言，覺得會分散注意力的話，不要顯示全拼提示是不是更好？真的有必要看提示嗎。

候選的完整讀音，RIME 只用來對顯示的編碼切分音節，畢竟編碼應該反應用戶的真實輸入，不然會無法理解退格鍵會刪除什麼；並且編碼除了切分不同之外，要與以下查出來的每個候選詞對應，而簡拼對應的每個候選字的全拼不盡相同。
要完整拼音提示的話，RIME 另有一個選項可以給候選詞標註完整拼音，是給每個候選詞單獨標註的。

變大寫作爲中間形式，這是大寫字母爲閒置字符時可以取巧的簡便做法。拼寫運算非常自由，當然可以用其他表示方法，比如不用一個大寫字母，而用與現有拼式不衝突的多個字符來表示中間編碼。就比如數學上字母不夠用，不是非得借用希臘字母，也可以寫 `x'` 嘛。

**Comment by byzod** [comment]
> 恕我直言，覺得會分散注意力的話，不要顯示全拼提示是不是更好？真的有必要看提示嗎。

如果只能显示一种的话，窝会显示全拼。以我的理解，双拼是一种输入编码方案，而不是输出编码方案，目前优化的方向也是基于键位来进行的。也就是说，以`fh [fang]`为例，`h`只是代表`ang`的键入编码，可以换成`ㄤ`，也可以换成`あ`，甚至`😅`，只要你键盘上有。实际上看着的`h`没有实际意义，只是代表`ang`，关注的应该是`ang`本身

另外窝是肌肉记忆派，打起来很6，但是给我看一堆jkuvalsjixz窝是不太看得懂的。需要原码主要是为了直出字符的情况，例如英语

> 畢竟編碼應該反應用戶的真實輸入，不然會無法理解退格鍵會刪除什麼

对啊所以双码显示也是以原始码为主的。当然有的输入法还 可以自由设置
 
> 而用與現有拼式不衝突的多個字符來表示中間編碼

首先还是会遇到同样的编码问题，xform是不支持非ascii编码的；其次由于除了`\U` `\L`以外我好像没找到其他可以自动批量替换字符的regex语法，由于regex只有单对单替换，而xlit又不能指定生效范围，要手替的话就只能写26行替换表达式……很蛋疼


要是输入码的部分也像其他地方一样有comment区域就好勒，就会方便操作很

而且就像rime-ice[配置里的说明](https://github.com/iDvel/rime-ice/blob/563ef2645ba995c45ba109dde60f7ac64a7443b9/double_pinyin_flypy.schema.yaml#L164)一样，这样做也会影响shift enter上屏的内容。如果可以打`fh [fang]`时enter上屏打`fh`，shift enter打`fang`就完美勒（现在是会显示处理后的结果`fh [fang]`

另一方便xform正则太多了性能也堪忧，现在打大概40个码就开始卡了……
